Backend-agnostic testing

ADBond opened the pull request:
This PR aims to provide a framework to easily allow backend-agnostic testing, so that it is straightforward to write tests that can run against all SQL backends. This currently covers `duckdb`, `spark`, and `sqlite` only.

The core of this provides a decorator `mark_with_dialects_excluding` which can be used to decorate test functions, which marks test with all dialects _except_ those specified, and any _groups_ those dialects belong to (see below). There is also an inclusive version `mark_with_dialects_including` which should be used only for specific dialect tests.

When tests that are marked in this way run, they receive a fixture `test_helpers` and a parameter `dialect`, which allows the use of an object `helper = test_helpers[dialect]` which has methods and properties to cover any dialect-specific part of testing, such as `helper.linker`, `helper.cl` (comparison library), and `helper.load_frame_from_csv()`.

Dialects can belong to zero or more groups - currently the only group is the privileged group `default` containing `spark` and `duckdb` which runs if no marks are passed in the command-line when running tests. These groups are just shortcuts to run sets of backends together.

There are a bunch of different options to run tests depending on which tests are to run. In the following `core` tests are precisely those which are not decorated with one of the new decorators (and should ultimately end up as more unit-testy stuff that does not care about backend, such as `test_input_column`,  `test_lat_long_distance`:
* `pytest -m core` - runs the `core` tests only
* `pytest -m spark` - runs the `core` tests and all `spark` tests (&sim for other backends)
* `pytest -m spark_only` - runs only `spark` tests (&sim for other backends)
* `pytest` - equivalent to `pytest -m default` - runs `core` tests, and those in the `default` group (`spark` and `duckdb`)
* `pytest -m all` - runs everything (`core` and all covered backends)

I have updated only a few test scripts with this approach, just to show how this works.

A couple of notes (which hopefully help sell the benefit of this approach for coverage):
* initially updated `tests/test_charts.py`, but one of the tests fails for `spark` due to a bug in handling of infinity in Bayes factors, so will leave that to a separate PR
* `test_estimate_prob_two_rr_match.py::test_prob_rr_valid_range` and `tests/test_u_train.py::test_u_train_link_only_sample` both fail for `sqlite`, seemingly due to separate bugs. This backend won't run for now in CI (which I haven't touched), so figure this is okay to go as-is. Will open separate issues for these

ADBond commented:
Have now added some docs to explain how these tests work, both from the perspective of writing and running

ADBond commented:
Have now added some docs to explain how these tests work, both from the perspective of writing and running

ThomasHepworth commented on docs/dev_guides/changing_splink/testing.md:
````diff
@@ -0,0 +1,283 @@
+
+Then we define the function, and pass parameters:
+
+```py linenums="3" hl_lines="1"
````
Oh cool, I did not know you could highlight lines

ThomasHepworth commented on docs/dev_guides/changing_splink/testing.md:
```diff
@@ -0,0 +1,283 @@
+
+### Tests in CI
+
+Splink utilises [github actions](https://docs.github.com/en/actions) to run tests for each pull request. This consists of a few independent checks:
+
+* The full test suite is run separately against several different python versions
+* The [example notebooks](./examples_index.html) are checked to ensure they run without error
+* The [tutorial notebooks](./demos/00_Tutorial_Introduction.html) are checked to ensure they run without error
```
These docs <img src="https://media.tenor.com/f9ifxTjrsTwAAAAj/chef-kiss.gif" width="40" height="40" />

ThomasHepworth commented on tests/conftest.py:
```diff
@@ -35,3 +50,30 @@ def df_spark(spark):
+# workaround as you can't pass fixtures as param arguments in base pytest
+# see e.g. https://stackoverflow.com/a/42400786/11811947
+@pytest.fixture
+def test_helpers(duckdb_helper, spark_helper, sqlite_helper):
+    return {
+        "duckdb": duckdb_helper,
+        "spark": spark_helper,
+        "sqlite": sqlite_helper,
+    }
```
Unless we plan on using `duckdb_helper`, `spark_helper`, etc as their own independent fixtures, I'd be tempted to rm them and yield:
```
    return {
        "duckdb": DuckDBTestHelper(),
        "spark": SparkTestHelper(spark),
        "sqlite": SQLiteTestHelper(),
    }
```

ThomasHepworth replied:
This is primarily to mitigate confusion

ADBond replied:
Yep I think you are right - think this structure is probably a hangover from trying to get the parameterisation to select the fixtures directly

ADBond replied:
Have made this change

ThomasHepworth commented on tests/helpers.py:
```diff
@@ -0,0 +1,129 @@
+        spark_frame.persist()
+        return spark_frame
+
+    def load_frame_from_csv(self, path):
+        df = self.spark.read.csv(path, header=True)
+        df.persist()
+        return df
```
We might as well add a `load_frame_from_parquet` method here too.

ADBond replied:
added

ThomasHepworth reviewed (commented):
I'll take another look over this again on Monday, but in general I think this is fantastic

ThomasHepworth commented on pyproject.toml:
```diff
@@ -62,3 +62,20 @@ ignore = [
+    "default",
+    "all",
+# backend-specific sets
+    "duckdb",
+    "duckdb_only",
```
I take it that `duckdb` and `duckdb_only` are identical?

Can you add a quick note

ThomasHepworth commented on tests/helpers.py:
```diff
@@ -0,0 +1,140 @@
+    def convert_frame(self, df):
+        return df
+
+    @property
+    def cll(self):
+        return cll_duckdb
+
+    @property
+    def cl(self):
+        return cl_duckdb
```
`ctl` too?

ADBond replied:
added

ThomasHepworth commented on pyproject.toml:
```diff
@@ -62,3 +62,20 @@ ignore = [
+# only tests where backend is irrelevant:
+    "core",
+# see tests/decorator.py::dialect_groups for group details:
+    "default",
```
Do we want `default` to run all of the benchmarks, or do we want this to be its own test mark?

ADBond replied:
I've taken benchmarking to be a separate workstream and consider them running separately, as they are performing different functions. As far as this PR goes I have just included the changes so that benchmarking runs the same as previously.
Might be a good upgrade to do something similar for benchmarking, but that might want some additional things (like slow/not slow type markers) and maybe a think about how we are doing the benchmark stuff

ADBond commented on pyproject.toml:
```diff
@@ -62,3 +62,20 @@ ignore = [
+addopts = ["-m default"]
+markers = [
+# only tests where backend is irrelevant:
+    "core",
```
as in are you doing just e.g. `pytest -m core` rather than `pytest -m core tests/`?

If so would you expect that to run the tests and benchmarking? If so I can either copy the functionality to benchmarking, or even perhaps lift `conftest` to root level so that it covers both

RossKen commented on docs/dev_guides/changing_splink/testing.md:
```diff
@@ -0,0 +1,283 @@
+* 'Core' tests - these are tests which test some specific bit of functionality which does not depend on any specific SQL dialect. They are usually unit tests - examples are testing [`InputColumn`](https://github.com/moj-analytical-services/splink/blob/master/tests/test_input_column.py) and testing the [latitude-longitude distance calculation](https://github.com/moj-analytical-services/splink/blob ...
+* Tests for specific backends - these are tests which run against a specific SQL backend, and test some feature peculiar to this backend. There are not many of these, as Splink is designed to run very similarly independent of the backend used.
+* Backend-agnostic tests - these are tests which run against some SQL backend, but which are written in such a way that they can run against many backends by making use of the [backend-agnostic testing framework](#backend-agnostic-testing). The majority of tests are of this type.
+
```
Do we need a caveat here/somewhere else to explain why Athena isn't included as a backend to test? I.e. because of needing a connection to an AWS account.

RossKen commented on docs/dev_guides/changing_splink/testing.md:
````diff
@@ -0,0 +1,283 @@
+
+### Running tests with docker 🐳
+
+If you want to test Splink against a specific version of python, the easiest method is to utilise docker 🐳.
+
+Docker allows you to more quickly and easily install a specific version of python and run the existing test library against it.
+
+This is particularly useful if you're using py > 3.9.10 (which is currently in use in our tests github action) and need to run a secondary set of tests.
+
+A pre-built Dockerfile for running tests against python version 3.9.10 can be located within [scripts/run_tests.Dockerfile](https://github.com/moj-analytical-services/splink/blob/master/scripts/run_tests.Dockerfile).
+
+To run, simply use the following docker command from within a terminal and the root folder of a splink clone:
+```sh
+docker build -t run_tests:testing -f scripts/run_tests.Dockerfile . && docker run --rm --name splink-test run_tests:testing
+```
+
+This will both build and run the tests library.
+
+Feel free to replace `run_tests:testing` with an image name and tag you're happy with.
+
+Reusing the same image and tag will overwrite your existing image.
+
+You can also overwrite the default `CMD` if you want a different set of `pytest` command-line options, for example
+```sh
+docker run --rm --name splink-test run_tests:testing pytest -W ignore -m spark tests/test_u_train.py
+```
````
Similar to other sections - I don't think this will apply to the majority of people, So I would be tempted to wrap up in a [collapsible block](https://squidfunk.github.io/mkdocs-material/reference/admonitions/#collapsible-blocks)

RossKen reviewed (commented):
Excellent work all round @ADBond! 🎉

The code all works well, the only comments I have left are on the docs to try and make them the most readable they can. This isn't necessarily a trivial structure so I really want to make sure it is as easy as possible to understand. Most of them are just suggestions so happy to discuss any to try and get to the best solution!

ADBond commented on docs/dev_guides/changing_splink/testing.md:
```diff
@@ -0,0 +1,283 @@
+
+Note that unlike the exclusive `mark_with_dialects_excluding`, this decorator will _not_ paramaterise the test with the `dialect` argument. This is because usage of the _inclusive_ form is largely designed for single-dialect tests. If you wish to override this behaviour and parameterise the test you can use the argument `pass_dialect`, for example `@mark_with_dialects_including("spark", "sqlite", ...
+
+## Running tests
```
Yeah I think that makes sense - running definitely going to be the more common case, and having a look back over it there isn't really any context from writing needed to understand how to run them

ADBond commented on docs/dev_guides/changing_splink/testing.md:
````diff
@@ -0,0 +1,283 @@
+```sh
+pytest tests/test_u_train.py::test_u_train_multilink
+```
+There may be many warnings emitted, for instance by library dependencies, cluttering your output in which case you can use `--disable-pytest-warnings` or `-W ignore` so that these will not be displayed. Some additional command-line options that may be useful:
+
+* `-s` to disable output capture, so that test output is displayed in the terminal in all cases
+* `-v` for verbose mode, where each test instance will be displayed on a separate line with status
+* `-q` for quiet mode, where output is extremely minimal
+* `-x` to fail on first error/failure rather than continuing to run all selected tests
+    * 
+* `-m some_mark` run only those tests marked with `some_mark` - see [below](#selecting-sets-of-tests) for useful options here
+
+For instance usage might be:
+```sh
+# ignore warnings, display output
+pytest -W ignore -s tests/
+```
+
+or
+```sh
+# ignore warnings, verbose output, fail on first error/failure
+pytest -W ignore -v -x tests/
+```
+
+You can find a host of other available options using pytest's in-built help:
+```sh
+pytest -h
````
Love these admonitions blocks - think they will prove great for breaking down docs information more easily

ADBond commented on docs/dev_guides/changing_splink/testing.md:
```diff
@@ -0,0 +1,283 @@
+
+The available options are:
+
+* `pytest tests/ -m core` - run only the 'core' tests, meaning those without dialect-dependence. In practice this means any test that hasn't been decorated using `mark_with_dialects_excluding` or `mark_with_dialects_including`.
+* `pytest tests/ -m duckdb` - run all `duckdb` tests, and all `core` tests
+    * & similarly for other dialects
+* `pytest tests/ -m duckdb_only` - run all `duckdb` tests only, and _not_ the `core` tests
+    * & similarly for other dialects
+* `pytest tests/ -m default` or equivalently `pytest tests/` - run all tests in the `default` group. The `default` group consists of the `core` tests, and those dialects in the `default` group - currently `spark` and `duckdb`.
+    * Other groups of dialects can be added and will similarly run with `pytest tests/ -m new_dialect_group`. Dialects within the current scope of testing and the groups they belong to are defined in the `dialect_groups` dictionary in [tests/decorator.py](https://github.com/moj-analytical-services/splink/blob/master/tests/decorator.py)
+* `pytest tests/ -m all` run all tests for all available dialects
```
Let me know if you have any thoughts/disagree on that

RossKen replied:
Yea I know it doesn't map on 1-1 but I think the headings are an improvement. Can we swap the second and third heading section so we follow the same core, backend agnostic, backend specific ordering that we have elsewhere?

RossKen replied:
Actually re-reading it, I think it is fine as is. Ignore me!

ADBond commented:
@RossKen I have made a bunch of changes to the docs page. I think I have addressed all of your suggestions but let me know if I've missed anything. Also let me know if I've done anything that wasn't what you had in mind/any further feedback

RossKen commented:
Thanks @ADBond, on my phone now but will look at properly in the morning

RossKen commented:
One naming thing. Would it make sense to do "Tests for specific backends" -> "Backend-Specific Tests" to better contrast with "Backend-Agnostic Tests"?

RossKen reviewed (approved):
As mentioned before - this is fantastic!

Thanks for making those changes to the docs - now even an idiot like me should be able to understand what is going on.

Good to merge from my end 🎉
